tests/static_map/custom_type_test.cu

```diff
@@ -89,10 +89,10 @@ struct hash_custom_key {
 
 // User-defined device key equality
 struct custom_key_equals {
-  template <typename custom_type>
-  __device__ bool operator()(custom_type lhs, custom_type rhs)
+  template <typename lhs_type, typename rhs_type>
```

Comment:
This is not enough yet. For the places that call `map.contains`, we should create a new key array of different type (but copied the same content of the original keys array). The new key type share implementation with the old type. So we may have something similar to:
```
template<typename T>
struct key_pair_type_base {.....};

template<typename T>
struct key_pair_type_lhs : struct key_pair_type_base {};

template<typename T>
struct key_pair_type_rhs : struct key_pair_type_base {};
```

Reply:
In addition, the test that calls `map.find` also makes me wonder if we also need a template key version for `find()`?

Reply:
> In addition, the test that calls `map.find` makes me wonder if we also need a template key version for `find()`?

I would prefer to make this PR dedicated to solving issue #172 unless template `find` is also a blocker for your current work. Fully supporting heterogeneous lookups #26 will be a separate PR that addresses all lookup functions.

Reply:
@ttnghia I think the new test should be sufficient. Please confirm.